File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
 
string solution(int n, int t, int m, vector<string> timetable) 
{
    vector<int> Crew;
    for (int i = 0; i < timetable.size(); i++)
    {
        string S_Hour = "";
        S_Hour = S_Hour + timetable[i][0];
        S_Hour = S_Hour + timetable[i][1];
        int Hour = stoi(S_Hour);
        
        string S_Minute = "";
        S_Minute = S_Minute + timetable[i][3];
        S_Minute = S_Minute + timetable[i][4];
        int Minute = stoi(S_Minute);
        
        int Time = Hour * 60 + Minute;
        Crew.push_back(Time);
    }
    sort(Crew.begin(), Crew.end());
 
    int Shuttle_Time = 540;
    int Crew_Idx = 0;
    int Answer_Time;
    for (int i = 1; i <= n; i++, Shuttle_Time = Shuttle_Time + t)
    {
        int Cnt = 0;
        for (int j = Crew_Idx; j < Crew.size(); j++)
        {
            if (Crew[j] <= Shuttle_Time)
            {
                Crew_Idx++;
                Cnt++;
                if (Cnt == m) break;
            }
        }
 
        if (i == n)
        {
            if (Cnt == m) Answer_Time = Crew[Crew_Idx - 1] - 1;
            else Answer_Time = Shuttle_Time;
        }
    }
    
    string answer = "";
    int Hour = Answer_Time / 60;
    int Minute = Answer_Time % 60;
    char A = Hour / 10 + '0';
    char B = Hour % 10 + '0';
    char C = Minute / 10 + '0';
    char D = Minute % 10 + '0';
    answer = answer + A;
    answer = answer + B;
    answer = answer + ':';
    answer = answer + C;
    answer = answer + D;
    return answer;
}

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.java
import java.util.*;

class Solution {
    public String solution(int n, int t, int m, String[] timetable) {
        
        // 가능한 가장 마지막 시간을 들고 있는다?
        int posTime = 0;
        Arrays.sort(timetable);
        
        int[] times = Arrays.stream(timetable)
            .mapToInt(time -> Integer.parseInt(time.replace(":", "")))
            .toArray();
        
        int qIdx = 0;
        int curTime = 900;
        
        for(int i=0; i<n; i++){
            int cnt = 0;
            
            for(int j=qIdx; j<times.length; j++){  
                if(cnt == m) break;
                
                if(times[j] <= curTime){
                    cnt++;
                    qIdx++;
                }
            }
            
            // 꽉 차는 경우 -> 마지막 친구의 -1분
            if(cnt == m){
                posTime = times[qIdx-1] - 1;
            }
            
            // 빈 자리가 있는 경우 -> 버스 도착 시간
            if(cnt < m){
                posTime = curTime;
            }
            
            // 분 단위 시간 컨트롤
            if(posTime%100 >= 60){
                posTime -= 40;
            }
            
            curTime += t;
            if(curTime%100 >= 60){
                curTime -= 60;
                curTime += 100;
            }
        }
        
        String ans = String.valueOf(posTime);
        
        if(posTime < 10){
            ans = "000" + ans;
        }
        else if(posTime < 100){
            ans = "00" + ans;
        }
        else if(posTime < 1000){
            ans = "0" + ans;
        }
        
        String answer = ans.substring(0,2) + ":" + ans.substring(2,4);
        
        return answer;
    }
}